numpy/polynomial/_polybase.py

```diff
@@ -105,7 +106,7 @@ class ABCPolyBase(abc.ABC):
     # the full set of superscripts and subscripts, including common/default
     # fonts in Windows shells/terminals. Therefore, default to ascii-only
     # printing on windows.
-    _use_unicode = not os.name == 'nt'
+    _use_unicode = os.name != 'nt' or 'ipykernel' in sys.modules
```

Comment:
Won't this also detect ipython used via a console?

Reply:
@eric-wieser `ipykernel` is specific to jupyter.  I've tested it on a couple of python versions (3.7, 3.8, 3.10) machines (windows 7, windows 10) and jupyers (notebook and lab). 
```
C:\Documents>ipython
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sys

In [2]: 'ipykernel' in sys.modules
Out[2]: False
```
Another (not 100% reliable but still) argument is 25 upvotes and 0 downvotes to this SO answer.

Reply:
Jupyter servers can be connected to from the terminal too though

Reply:
@eric-wieser You mean jupyter-console? Its description says "An IPython-like terminal frontend for Jupyter kernels in any language." So its primary use case is to provide a functionality similar to ipython to other languages (julia, rust, etc). Running jupyter-console for python is an exceptionally narrow use case. The only reason to do it that I could think of is connecting to an already running jupyter session for debugging. But you need to love console to prefer it over connecting to it via another graphical jupyter notebook. And people who love console know how to make it support unicode :) Actually, Unicode problems on Windows are in the past nowadays. They were relevant maybe 5-7 years ago, but even then there were third-party tools to fix them. 

The vast majority of users are running jupyter from the browser (or a browser-like environment like Microsoft Visual Code). And it is strange to see ascii fallback there.

Reply:
> The vast majority of users are running jupyter from the browser

I wonder if there is any data around to prove this. I have seen corporate workflows where users are connecting to remote kernels via different terminal-like IDEs and consoles. I don't know if it is a small minority or a significant minority or maybe even a majority of users, but even 1% of the scientific python users is a significant number of people who might be affected by this.

It is hard to judge the impact of this change. My guess is that even on windows most terminals and consoles now support unicode, so this should be safe. Maybe we should go further and make unicode the default everywhere, in order to get more feedback about where this could break `str` representation.

Reply:
@mattip What is the oldest Windows version that NumPy officially supports now? 

I've checked Windows Server 2003 (a server version of Windows XP). It does not support Unicode in the terminal out of the box, but you can't install modern Python on it, either. Python 3.4 is the most recent version I managed to install there.

Reply:
NumPy supports 3.8 and upwards

Reply:
> The vast majority of users are running jupyter from the browser (or a browser-like environment like Microsoft Visual Code). And it is strange to see ascii fallback there.

Won't these users see the LaTeX formatter anyway?

Reply:
While I suspect this works at pretty well in practice, I would tend towards what Eric said:  Notebook users can use the pretty latex representation which is much nicer anyway?!

Reply:
`'ipykernel' in sys.modules` also `True` when working with [Spyder](https://www.spyder-ide.org/). This might not be an issue, probably unicode works fine from spyder (working on linux now, so I cannot verrify).

Reply:
> Unicode is a natural choice in all of them.

I don't disagree with this, I just don't think `ipykernel` offers a meaningful distinction for whether this behavior is desired. If I want unicode in my log files, I almost certainly don't care whether or not those logs were produced by an interpreter running through Jupyter.

Probably the way to move forward here is expose some public API for users to turn off unicode printing, and change it to be on by default universally.

Reply:
> > The vast majority of users are running jupyter from the browser (or a browser-like environment like Microsoft Visual Code). And it is strange to see ascii fallback there.
> 
> Won't these users see the LaTeX formatter anyway?

Actually, what users see in python and ipython console is also not `str`. They see `repr`:

``` python
In [1]: import numpy.polynomial as poly

In [2]: p = poly.Polynomial([1/7, 1/3])

In [3]: p
Out[3]: Polynomial([0.14285714, 0.33333333], domain=[-1,  1], window=[-1,  1], symbol='x')
```

So strictly speaking `str` has no way of getting into console 'by mistake'. Only by deliberately converting with something like `str(p)` or `f'{p}'` and printing it.

Also I did some more experiments on Windows 7 and there are still problems with printing degrees higher than 3 (they are displayed with empty squares) to console.

Maybe it would make sense to have `p.to_ascii()` which would always return ascii and `str(p)` which would always (=on any OS) return unicode?